### DISCUSSION TITLE
KSTAR Modbus Protocol V3.5 (SN: 23XXXXXXXX) new Documentation from KSTAR

### DISCUSSION THREAD

**Discussion by Schockolade** [StephanJoubert/home_assistant_solarman#543] [Ideas]
Hi guys,

I am currently trying to get the data from two KSTAR BluE-6KT-M1 (SN: 23XXXXXXXXXX, module version LSW3_15_1214_1.20) inverters into the Homeassistant.

Since the previous ways did not really work (e.g.: kstar_hybrid.yaml which is based on the V2.5 of the protocol) because the registers do not seem to fit, I asked KSTAR for a current documentation. There I received the attached PDF. 

The statement from KSTAR:
"Hello, the protocol you are using [V2.5] is for energy storage products, so it does not correspond to the data of your pure grid connected inverter. You can use the protocol in the attachment"

With the new PDF, I have entered the most important values for me in the custom_parameter.yaml as I understand the PDF and this add-on (also attached).

After a lot of trial and error and moving registers back and forth, I was able to adjust the custom_properties.yaml so that the values fit. The PDF from KASTR was very helpful.
If you have similar inverters or the kstar_hybrid.yaml does not work you are welcome to try the custom_properties.yaml.

I also noticed that, according to the documentation, some registers contain U8 values. This doesn't seem to be covered by the integration, does it? Would this be worth expanding?

I hope I can help someone with the V3.5 documentation and further informations.

[1_KSG1－250K Inverter Modbus Communication  Protocol 3.5.pdf](https://github.com/StephanJoubert/home_assistant_solarman/files/14762119/1_KSG1.250K.Inverter.Modbus.Communication.Protocol.3.5.pdf)

Updated .yaml File:
[custom_parameters.yaml.txt](https://github.com/StephanJoubert/home_assistant_solarman/files/14776512/custom_parameters.yaml.txt)

**Comment by frsantos** [comment]
> Hi guys,
Hi @Schockolade 

> I am currently trying to get the data from two KSTAR BluE-6KT-M1 (SN: 23XXXXXXXXXX, module version LSW3_15_1214_1.20) inverters into the Homeassistant.
>
> Since the previous ways did not really work (e.g.: kstar_hybrid.yaml which is based on the V2.5 of the protocol) because the registers do not seem to fit, I asked KSTAR for a current documentation. There I received the attached PDF.

Nice

> The statement from KSTAR:
"Hello, the protocol you are using [V2.5] is for energy storage products, so it does not correspond to the data of your pure grid connected inverter. You can use the protocol in the attachment"

The definition file I created was for KSTAR hybrid inverters (connected to a battery or other energy sources), while yours seems to be a normal inverter, right? I suppose you can create a PR with a new definition file for this KStar inverter, same as there are different definition files  with different registers for DEYE or SOFAR inverters.

> With the new PDF, I have entered the most important values for me in the custom_parameter.yaml as I understand the PDF and this add-on (also attached).
>
> After a lot of trial and error and moving registers back and forth, I was able to adjust the custom_properties.yaml so that the values fit.

That is nice. 

Remember that `custom_properties.yaml` file is only for your local use, and please keep a back up of it. I remember it being deleted when the integration was updated.

> I also noticed that, according to the documentation, some registers contain U8 values. This doesn't seem to be covered by the integration, does it? Would this be worth expanding?

The U8 is just a byte.  The hybrid definition file reads some U8, for example as a text as a bunch of chars ("Inverter SN Number", registers 3228-3238, rule 5) or a number from a byte ("ARM Version Number", register 3216, rule 1). The problem with the bytes is what kind of information are held in them, and that is up to us to process. 

For example, the ARM/DSP version number ("VX.Y.Z") are set in the two bytes on each register. The first byte contains the
 X.Y part, and the second byte contains the Z part. That needs some math that the integration is not capable of doing, so probably a HA helper is needed for that transformation.
